[docs] Add more package templates based on Conan v2 with support to Conan v1

uilianries opened the pull request:
After #12575

- Add cross-compatible Conan v1 & v2 template based on header-only example
- Add cross-compatible Conan v1 & v2 template based on autotools example
- Add cross-compatible Conan v1 & v2 template based on msbuild example
- Add cross-compatible Conan v1 & v2 template based on pre-built example
- Move CMake example to new place and fix its usage according Conan v2

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

prince-chrismc commented on docs/package_templates/autotools_package/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+cmake_minimum_required(VERSION 3.8)
+
+project(test_package C) # if the project is pure C
```
since it's a CPP file in this example can we comment this line out? this way its more usable

```suggestion
#project(test_package C) # if the project is pure C
```

uilianries replied:
I just converted the autotools test_package to a C usage. Now we have a separated example, not only c++.

prince-chrismc commented on docs/package_templates/autotools_package/all/test_package/test_package.cpp:
```diff
@@ -0,0 +1,16 @@
+#include <cstdlib>
+#include <iostream>
+#include "package/foobar.hpp"
```
```suggestion
#include "package/foobar.hpp" // Make sure includes work as expected
```

uilianries replied:
added, done.

prince-chrismc commented on docs/package_templates/autotools_package/all/test_package/test_package.cpp:
```diff
@@ -0,0 +1,16 @@
+    std::cout << "Avoid background apps or servers." << std::endl;
+    std::cout << "The propose is testing the generated artifacts only." << std::endl;
+
+    foobar.print_version();
```
```suggestion
    foobar.print_version(); // Make sure to call something that will require linkage for compiled libraries
```

uilianries replied:
added, done.

prince-chrismc commented on docs/package_templates/autotools_package/all/test_v1_package/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+cmake_minimum_required(VERSION 3.8)
+
+project(test_package C) # if the project is pure C
```
```suggestion
#project(test_package C) # if the project is pure C
```

prince-chrismc commented on docs/package_templates/cmake_package/all/conanfile.py:
```diff
@@ -0,0 +1,167 @@
+        tc.cache_variables["PACKAGE_CUSTOM_DEFINITION"] = True
+        if is_msvc(self):
+            # don't use self.settings.compiler.runtime
+            tc.cache_variables.definitions["USE_MSVC_RUNTIME_LIBRARY_DLL"] = "MD" in msvc_runtime_flag(self)
```
There's an "is static runtime" helper as well

prince-chrismc reviewed (approved):
Super minor suggestions

uilianries commented:
@prince-chrismc Thank you for reviewing, I just updated according to your review, but I changed the autotools test_package to a C project, so we have both C++ and C cases as examples.

uilianries commented:
@prince-chrismc I did an update:

- Removed unused import from msbuild example: https://github.com/conan-io/conan-center-index/pull/12678/commits/3923970d511011af55c31f82a131a0a548262f39
- After doing some tests here, `apply_conandata_patches` can be executed under `build()` for header-only packages, including when `no_copy_source` is declared: https://github.com/conan-io/conan-center-index/pull/12678/commits/a4b69cf75b7f6f866bf59ebb1e2db24303aadc97
- Add a new template for pre-built binaries: https://github.com/conan-io/conan-center-index/pull/12678/commits/1750db8a7164bdd2daf1ba6cb3af050a99f9c453

uilianries commented:
Maybe I should close this PR and open separated PRs, one for each example. It reached 800 lines, no one wants to review it.

SpaceIm commented on docs/package_templates/header_only/all/conanfile.py:
```diff
@@ -0,0 +1,117 @@
+        # src_folder must use the same source folder name the project
+        basic_layout(self, src_folder="src")
+
+    # not mandatory when there is no patch, but will suppress warning message about missing build() method
+    def build(self):
+        # apply patches in source_folder
+        apply_conandata_patches(self)
```
seems wrong where there is `no_copy_source = True`

uilianries replied:
> I would advice to remove no_copy_source = True when patching is required.

Not bad idea at all. 

Indeed, when patching under the build() method, the patch will be applied over source_folder, so it works, however, it looks wrong. I gonna update it and follow your idea.

SpaceIm commented on docs/package_templates/header_only/all/conanfile.py:
```diff
@@ -0,0 +1,117 @@
+        get(**self.conan_data["sources"][self.version],
+            destination=self.source_folder, strip_root=True)
+
+    def validate(self):
+        # validate the minimum cpp standard supported
+        if self.info.settings.compiler.cppstd:
+            check_min_cppstd(self, self._minimum_cpp_standard)
+        minimum_version = self._compilers_minimum_version.get(str(self.info.settings.compiler), False)
+        if minimum_version and Version(self.info.settings.compiler.version) < minimum_version:
+            raise ConanInvalidConfiguration(f"{self.name} requires C++{self._minimum_cpp_standard}, which your compiler does not support.")
+        # in case it does not work in another configuration, it should validated here too
+        if self.info.settings.os == "Windows":
+            raise ConanInvalidConfiguration(f"{self.ref} can not be used on Windows.")
```
move this before build_requirements()

uilianries replied:
Talking about it, I just opened an issue to update conan create methods order.

SpaceIm commented on docs/package_templates/header_only/all/conanfile.py:
```diff
@@ -0,0 +1,117 @@
+        if self.info.settings.os == "Windows":
+            raise ConanInvalidConfiguration(f"{self.ref} can not be used on Windows.")
+
+    def layout(self):
+        # src_folder must use the same source folder name the project
+        basic_layout(self, src_folder="src")
```
before source()?

uilianries replied:
Yes, gonna update.

SpaceIm commented on docs/package_templates/header_only/all/conanfile.py:
```diff
@@ -0,0 +1,117 @@
+        for p in self.conan_data.get("patches", {}).get(self.version, []):
+            copy(self, p["patch_file"], self.recipe_folder, self.export_sources_folder)
+
+    # same package ID for any package
+    def package_id(self):
+        self.info.clear()
+
```
to remove, I see this method twice with different comments

uilianries replied:
Good catch! I saw it locally and forgot to remove, thank you!

SpaceIm commented on docs/package_templates/header_only/all/conanfile.py:
```diff
@@ -0,0 +1,117 @@
+
+    def validate(self):
+        # validate the minimum cpp standard supported
+        if self.info.settings.compiler.cppstd:
```
AFAIK, doesn't work yet if there `self.info.clear()` in `package_id()`

uilianries replied:
True, I'll remove. The point is, on Conan 2.0 cppstd will be always expected.

uilianries replied:
I've updated to `self.info.settings.get_safe("compiler.cppstd")`. compiler is None when building, but when installing, it will be used to validate.

SpaceIm commented on docs/package_templates/header_only/all/conanfile.py:
```diff
@@ -0,0 +1,117 @@
+            check_min_cppstd(self, self._minimum_cpp_standard)
+        minimum_version = self._compilers_minimum_version.get(str(self.info.settings.compiler), False)
+        if minimum_version and Version(self.info.settings.compiler.version) < minimum_version:
+            raise ConanInvalidConfiguration(f"{self.name} requires C++{self._minimum_cpp_standard}, which your compiler does not support.")
```
```suggestion
            raise ConanInvalidConfiguration(f"{self.name} {self.version} requires C++{self._minimum_cpp_standard}, which your compiler does not support.")
```

SpaceIm replied:
or maybe `self.ref`?

uilianries commented:
@SpaceIm thank you very much for reviewing. I did all updates

prince-chrismc commented:
> * After doing some tests here, `apply_conandata_patches` can be executed under `build()` for header-only packages, including when `no_copy_source` is declared:

Thats an amazing improvement!!! I did not know that was a thing 🚀

prince-chrismc commented on docs/package_templates/header_only/all/conanfile.py:
```diff
@@ -0,0 +1,114 @@
+
+    # not mandatory when there is no patch, but will suppress warning message about missing build() method
+    def build(self):
+        # apply patches in source_folder. The attribute no_copy_source should not be used when applying patches
```
```suggestion
```

uilianries replied:
My suggestion is keeping 'The attribute no_copy_source should not be used when applying patches'. So we can copy to the build folder, apply the patches and keep source folder immaculate.

SpaceIm replied:
Yes, very important to remove `no_copy_source = True` in this case, otherwise 2 successive builds through conan create/install will try to apply patches in source folder, so the first one will modify files under source folder, and the second build will fail during this new attempt to patch files.

prince-chrismc replied:
I didn't not notice the second half 

```suggestion
        # The attribute no_copy_source should not be used when applying patches in build
```

uilianries replied:
Done, applied last suggestion.

prince-chrismc commented on docs/package_templates/prebuilt_tool_package/all/conanfile.py:
```diff
@@ -0,0 +1,57 @@
+        copy(self, pattern="foo", dst=os.path.join(self.package_folder, "bin"), src=self.source_folder)
+
+    def package_info(self):
+        # folders not used for pre-built binaries
```
I saw a convo on slack about this, but might be worth mention the no layout here, try to define the min?

uilianries replied:
Good point. what about:

```python
def layout(self):
    pass
```

So we will not generate warning and will keep explicit to do nothing related to layout. WDYT?

prince-chrismc replied:
👍 that's perfect

uilianries replied:
done.

prince-chrismc commented:
> Maybe I should close this PR and open separated PRs, one for each example. It reached 800 lines, no one wants to review it.

That's a decent idea. I dont mind pushing this in and fixing in a follow up PRs.

it's 98% good which is better then what we have :)
